[RFC 0165] Bootspec v2

RaitoBezarius opened the pull request:
This introduces the second revision of RFC-0125, Bootspec, addressing the feedback we received in https://github.com/NixOS/rfcs/pull/125 and building on our experience of Bootspec v1.

[Rendered](https://github.com/RaitoBezarius/nixos-rfcs/blob/bootspecv2/rfcs/0165-bootspecv2.md)

TODO:

- [x] Clarify transition period: generate v1 and v2 plus offer synthesis to v1 using v2 and non-bootspec data
   - [x] Mention existing infrastructure tools that will take a NixOS closure and transform it into something else
- [x] Move initrd secrets feature inside its own extension
- [ ] Devicetree discussion
   - [ ] Do we collapse `FDTDIR` / `devicetree` into… nothing or just `FDTDIR`?

RaitoBezarius committed:
[RFC 0165] Bootspec v2

This introduces the second revision of RFC-0125, Bootspec,
addressing the feedback we received in https://github.com/NixOS/rfcs/pull/125
and building on our experience of Bootspec v1.

grahamc commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+    // (Required) List of paths to the initrd, can be empty
+    "initrds": [ "/nix/store/xxx-initrd-linux/initrd" ],
+
+    // (Required) Hash map of desired secrets for that generation inside of the initrd.
+    // Implementors of a bootloader installation procedure should examine their options
+    // to securely make available the secret inside the initrd phase.
+    // This may involve leveraging TPM2 via systemd-credentials or any measure you deem
+    // to be reasonable in the context.
+    // The legacy behavior is to prepare a CPIO archive for each file and
+    // extend the `initrds` fields with those CPIO archives.
+    // Make sure the location where the secrets are dropped in the initrd are visible
+    // for the user.
+    "initrdSecrets": {
+      "my-private-key": "/etc/nixos/secrets/wireguard-private-key",
+    }
```
I wonder if the various implementations possible with initrd secrets makes this design too specific for the design space. What if this mechanism was split out into an "extension" like specifications? I could imagine an initrdSecrets variant here which is pulling secrets from Vault, or AWS Secrets Manager, or `pass`.

RaitoBezarius replied:
I touch on the matter in the FAQ, I believe that it's up to the bootloader implementation to offer another backend for interpreting the values of the `initrdSecrets` if needed. Some trivial secret engine could even offer a filesystem to mount the secrets and let them be downloaded/offered at runtime.

RaitoBezarius commented:
I would like to particularly invite @samueldr for clear expertise on devicetree and non-x86 systems in Nix(OS) ecosystem (of course, you may not have time, and I am already thankful for the time you provided me online).

I would like @lheckemann to join me again as a co-author if that's possible.

I would like to invite the previous shepherds of bootspec: @govanify @06jellyjac.
I would like to invite the people who made a lot of relevant comments which prompted this version 2: @hesiod 

I would like to invite my colleagues of lanzaboote: @nikstur and @blitz to comment from their perspective as they worked a lot with Bootspec v1, maybe you can even become shepherd this time!

I would like to invite the initial authors of the previous revision: @grahamc and @cole-h — thank you for your support!

And I am probably forgetting a lot of folks, so everyone is welcome if you want to participate in this second revision.

This work is funded by the Tech Sovereign Fund as part of https://github.com/nix-community/projects/blob/main/proposals/nixpkgs-security.md#boot-chain-security (milestone 3), if someone wants to take a chunk of the money working on that with an active role, I would be glad to distribute some in the limits of classical bureaucracy and goal definitions regarding this specification.

cole-h commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+# Summary
+[summary]: #summary
+
+Bootspec v2 is the second revision of the Bootspec document, introduced in RFC-0125.
```
Nit: Would be swell to link directly to RFC 125's text at this mention.

grahamc commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+# Proposed Solution
+[proposed-solution]: #proposed-solution
+
+- `initrd` will be removed from the v2
+- `initrds` will be introduced as a list of initrd (compressed or uncompressed
```
Is the intent to generate both v1 and v2 bootspec documents for a transition period, or to do that in synthesis?

RaitoBezarius replied:
Good question, both options are valid IMHO. I guess that we can have something like:

- Always the latest generation is generated
- Synthesis can enrich a boot.json with a potentially old version, using information from the top-level or from new versions.

What do you think about that?

RaitoBezarius replied:
Let's do both as I don't think it's very expensive to do anyway

grahamc replied:
By infrastructure I mean services that accept a closure over the network and transforms the NixOS system into something else. That's why the forward/backward compat is useful, and was a key use case I wanted to support with bootspec.

RaitoBezarius replied:
This makes sense to me

cole-h commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+name used to organize secrets, and the value denotes an accessible path during
+bootloader installation.
+
+_Example_
```
nit:

```suggestion
### Example
```

makes it easier to discover examples in the github "go to headings" list

cole-h commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+### Milestones
+[milestones]: #milestones
+
+- Update Bootspec with the version 2 of that specification.
```
I imagine this includes both `bootspec`, the Rust library, and its interface in Nixpkgs?

RaitoBezarius replied:
Yep

grahamc commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+fetching at activation time, e.g. a FUSE secretfs that will dynamically query
+the secret engine for a secret and make it available for short time.
+
+If any logic requires user interaction, it is preferable to invest in a custom
+bootloader installer logic and use the static fields to refer to secrets that
+will be requested. We never defined "path" on purpose as it could be a network
+path or an application-specific path.
+
+It could even be a UEFI device path!
```
I don't think this is a good idea, and opens up a lot of risky potential security issues where a security-sensitive value may be mishandled. I think it would be much more preferable to have a strong definition of what these values are and how they are to be interpreted. I think it would be better if different behavior was implemented by extensions which also had strong definitions of how to handle the values.

Putting undefined at the trust root smells bad to me.

RaitoBezarius replied:
Hmm, so with your idea, would we be doing formalized extensions of `initrdSecrets` and let people experiment with their own definitions of `initrdSecrets` ?

RaitoBezarius replied:
And we would strongly say those are paths on the local system, right?

grahamc replied:
My preference would be `org.nixos.bootspec.v2` have no mention of `initrdSecrets`, and the definition l interface is in a separate extension like `org.nixos.initrd-secrets.v1` (which could be defined in this same RFC) or something.

RaitoBezarius replied:
OK, this makes sense to me.

RaitoBezarius replied:
I actually also like this idea

grahamc replied:
I was actually loathe to include initrd secrets in v1, but I don't think we'd thought about Extensions yet, and also didn't think to clean it up after extensions were added.

grahamc commented:
By the way, I like this RFC and I think its going in a good direction 👍 thank you!

infinisil commented on rfcs/0164-bootspecv2.md:
Why does this need to be an RFC? Nothing here stands out as being controversial or substantial. It seems like a normal PR would do the job.

RaitoBezarius replied:
Previous RFC showed that people had important and interesting comments to bring to reach to a good design. This RFC is done with the same effect in mind, I don't think we have controversial things, but we have substantial things like modifying the initrd secrets mechanism substantially in the future.

grahamc replied:
Defining the interface in a way that it is defined by RFCs, and adjusted by RFCs was outlined in the initial RFC for bootspec: https://github.com/NixOS/rfcs/blob/master/rfcs/0125-bootspec.md#goals

Since it is an API for internal and external consumption, the stability and careful iteration are key properties for bootspec to be successful at its purpose.

infinisil replied:
Makes sense. I don't think the RFCs repo should be used as the canonical place for the specification though, because RFCs are set in stone and not meant to be updated, even for minor changes.

RaitoBezarius replied:
This is not really an update per se, but a new specification. Bootspec v1 will continue to exist and be used in production.

grahamc replied:
Right. Bootspec revisions aren't updated, new versions are designed and agreed upon.

infinisil replied:
I think they _should_ be updateable, and people shouldn't have to refer to RFCs for specifications. Even if there's no typos or minor revisions necessary, just having both versions next to each other, or linking to the new spec from the old one would be very beneficial. With it just being in an RFC it's hard to compare it to the previous version for future readers without reading through all the additional argumentation in the RFC that isn't necessary for a specification.

And this doesn't mean you can't use RFCs to create new versions of specifications, it just means that you should decide a canonical location for it and commit it there once the RFC is accepted. Nix RFCs (contrary to IEFT RFCs) are only there to keep the historical context for why a decision was made, they shouldn't be needed to know what the decision was.

amjoseph-nixpkgs replied:
> Why does this need to be an RFC? Nothing here stands out as being controversial or substantial. It seems like a normal PR would do the job.

I agree, but somebody committed a strongly-worded comment to the repository which asserts otherwise:

https://github.com/NixOS/nixpkgs/blob/686a5fca36d6ebc0e2fda1c4fac1717e318c4de5/nixos/modules/system/activation/bootspec.nix#L2-L3

```nix
# This document is considered a stable API, and is depended upon by external tooling.
# Changes to the structure of the document, or the semantics of the values should go through an RFC.
```

I do not agree with this comment.  I would like to see it removed.

Note that this is a cue schema, and the bootspec machinery will reject any extensions which aren't part of the schema.

I understood going FCP for the RFC, without an official way of passing devicetrees, and without me making a loud objection to that, meant that there wouldn't be an official way of using devicetrees with bootspec, but that we wouldn't be blocked from experimenting with unofficial mechanisms.

I was pretty unhappy to discover that the implementation had decided to interpret the RFC as a binding agreement to *forbid devicetrees altogether*.  That felt like quite a bait-and-switch.  I would have made loud noises during FCP if this had been made clear.

RaitoBezarius replied:
@amjoseph-nixpkgs We are not blocked to change `bootspec.nix` to allow experimental extensions neither experimenting with unofficial mechanisms. No one approached me with any of that, though, so I was kinda stuck with my U-Boot experiments.

Extensions were completely formalized as part of the RFC, so anything was open!

We just want to go through RFC to formalize additional extensions we promised to get around and make them coherent for any userspace which may depend upon it.

grahamc replied:
> I think they _should_ be updateable, and people shouldn't have to refer to RFCs for specifications. Even if there's no typos or minor revisions necessary, just having both versions next to each other, or linking to the new spec from the old one would be very beneficial. With it just being in an RFC it's hard to compare it to the previous version for future readers without reading through all the additional argumentation in the RFC that isn't necessary for a specification.
> 
> And this doesn't mean you can't use RFCs to create new versions of specifications, it just means that you should decide a canonical location for it and commit it there once the RFC is accepted. Nix RFCs (contrary to IEFT RFCs) are only there to keep the historical context for why a decision was made, they shouldn't be needed to know what the decision was.

I emphatically disagree, though if you’d like to pursue amending something decided in an RFC (like where to track and amend the boot spec format) probably a new RFC is the way to go.

infinisil replied:
Actually this is in [the RFC process readme](https://github.com/NixOS/rfcs#the-rfc-life-cycle):

> RFC documents are intended to be seen as the documentation of a decision and a snapshot of a moment in time, rather than a specification-like normative document. Think more of a Matrix Spec Proposal and less like an IETF RFC. Therefore, once accepted, RFCs should generally not be substantially changed. Only very minor changes should be submitted as amendments (via a follow-up pull request). It is the general expectation that any information intended to be normative and "outlive" the initial RFC process should live outside of the RFC document, mostly in documentation and code. These may be subject to change as usual, and of course any "substantial" changes will again require a new RFC. Usually there is no need to update the original RFC to keep it up with updates on the implementation.

blaggacao replied:
Seems unambiguous enough to kindly ask the proponents to decide on another canonical location for the spec.

Let's not continue the exception.

RaitoBezarius replied:
I tried to address things in this RFC regarding moving the development of next iterations.

RaitoBezarius replied:
Let me know if this is insufficient, resolving in the meantime.

grahamc commented:
I emphatically disagree, though if you’d like to pursue amending something decided in an RFC (like where to track and amend the boot spec format) probably a new RFC is the way to go. 

On Wed, Nov 1, 2023, at 6:35 PM, Silvan Mosberger wrote:
> 
> 
> ***@***.**** commented on this pull request.
> 
> 
> On rfcs/0164-bootspecv2.md <https://github.com/NixOS/rfcs/pull/165#discussion_r1379398502>:
> 
> I think they *should* be updateable, and people shouldn't have to refer to RFCs for specifications. Even if there's no typos or minor revisions necessary, just having both versions next to each other, or linking to the new spec from the old one would be very beneficial. With it just being in an RFC it's hard to compare it to the previous version for future readers without reading through all the additional argumentation in the RFC that isn't necessary for a specification.
> 
> And this doesn't mean you can't use RFCs to create new versions of specifications, it just means that you should decide a canonical location for it and commit it there once the RFC is accepted. Nix RFCs (contrary to IEFT RFCs) are only there to keep the historical context for why a decision was made, they shouldn't be needed to know what the decision was.
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/NixOS/rfcs/pull/165#discussion_r1379398502>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAASXLHBSMHV5PHGQIWEL23YCLFCLAVCNFSM6AAAAAA6ZOZO42VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMBZGE4DCMRXGE>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

amjoseph-nixpkgs commented:
> I emphatically disagree, though if you’d like to pursue amending something decided in an RFC (like where to track and amend the boot spec format) probably a new RFC is the way to go.

This is not how I understand the RFC process to work, nor how it describes itself:

https://github.com/NixOS/rfcs/blob/c8569f6719356009204133cd00d92010889ed56d/rfcs/0001-rfc-process.md?plain=1#L47-L49

RFCs are for "**substantial** changes" only.

Reinterpreting the RFC process as granting "write-protect unless RFC" for even minor changes would make the RFC process even more dysfunctional than it already is.

See also https://github.com/NixOS/rfcs/pull/165#discussion_r1379648995

amjoseph-nixpkgs commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+
+The latter indicates a potentially problematic, non-upstreamed, or in-development platform.
+
+The current Bootspec v1 does not formally encode information about hardcoded device trees or the folder containing available device trees. As a result, unformalized extensions are needed to address these fundamental use cases.
```
```suggestion
The current Bootspec v1 does not formally encode information about device trees or the folder containing available device trees. As a result, unformalized extensions are needed to address these fundamental use cases.
```

RaitoBezarius replied:
(I'm fine with applying this change once we finish the discussion on DTB/DTOs and we reach to a conclusion.)

amjoseph-nixpkgs commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+
+The current Bootspec v1 does not formally encode information about hardcoded device trees or the folder containing available device trees. As a result, unformalized extensions are needed to address these fundamental use cases.
+
+Regarding overlays in NixOS, any expressed overlay is incorporated into the final device tree. This eliminates the necessity to formalize an additional overlays field list, as overlays can be transformed into device trees as needed.
```
See earlier you complained about devicetrees being hardcoded, and then here you say "btw we think non-hardcoded devicetrees are lame so we won't allow them".  It's circular reasoning...

If you want to keep things simple and require one single `.dtb` file for bootspec v2 that's fine, since a single `dtb` is better than none at all.  Just please don't portray limitations of the simplifying assumption you're making as being limitations of devicetrees.

At the end of the day devicetrees are basically a cleaner version of ACPI without the halting problem or marketing budget.  They're data, not code.  They don't have bigtech employees running around trumpeting their virtues.  All the other differences are basically just syntax.

RaitoBezarius replied:
Well, foremost, DTOs are exactly implemented like this in NixOS.

I don't think they are lame, but it has been brought to me again by @samueldr that DTOs are not necessarily the panacea we have been looking for and it's unclear to me what would be the provided benefits of supporting a first-class field for DTOs inside Bootspec.

So please enlighten me:

> If you want to keep things simple and require one single .dtb file for bootspec v2 that's fine, since a single dtb is better than none at all. Just please don't portray limitations of the simplifying assumption you're making as being limitations of devicetrees.

Can you mention some examples of introduced limitations via our simplifying assumptions we are doing by bolting any DTOs inside DTBs ?

I had some ideas of my own: multiple-devices variants support, etc. But I am not convinced they are not supported with the current proposal.

Also, please get me right: I'm not saying they are lame, I'm saying we can re-encode any DTOs as a DTB, therefore, we don't need (until proven wrong?) DTOs as part of the generation data, but we should "erase" it via the NixOS module machinery.

samueldr replied:
Is there more than whining and FUD that you want to apply to this section, @amjoseph-nixpkgs?

Reminder that bootspec is about *listing the knowledge from a NixOS generation for configuring boot*.

There are no generic methods of describing overlays that will apply to generic boot methods.

Or maybe you can elucidate this whole situation and show what is the generic device tree overlay spec in practice, and what form of ***knowledge*** that is needed to expose in bootspec to *allow configuring the boot for a specific given generation*.

amjoseph-nixpkgs commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+
+Initrd secrets serve to protect boot-time secrets from exposure within the Nix
+store. This is achieved by using runtime scripts that append CPIO archives to
+the generation's initrd during bootloader installation. However, the term
+"initrd secrets" can be misleading, as the secrets are plaintext and can be
+accessed from the ESP partition, offering limited confidentiality and no
+integrity.
```
Apparently I have made this mistake.  I don't use NixOS often, but I do have two machines that run it.  The "initrd secrets" seemed like a good place to put my root volume encryption key (!)  Fortunately these machines have no ESP partition at all.  But yeah, that name is misleading.

I guess it the intended meaining is "*non-world-readable* portion of the `initrd`"?

RaitoBezarius replied:
Urgh, so sorry for that. Yeah, it's the "non-world-readable portion of `initrd`" from the Nix store perspective.

- It is still world-readable if you are an unprivileged process that can access the ESP.
- It is still world-readable if you boot another system that can access the ESP.

It is a very weak concept of "non-world-readability" where you only focus on protecting what ends up in the Nix store.

RaitoBezarius commented:
> Reinterpreting the RFC process as granting "write-protect unless RFC" to certain files in nixpkgs, covering even minor experimental changes, would make the RFC process even more dysfunctional than it already is.

We are not saying that the RFC process grants write protect unless RFC for minor experimental changes. The extension field offers full flexibility for anyone to do even **major experimental changes**, we just want to ensure that all the ecosystem is protected by strong stability properties, e.g. breaking changes are strictly forbidden, semantics change in the interpretation of values are highly ill-advised (re: the discussion on nulls).

Maybe what this is all hinting at is that we need another way to track the continuous life of specifications like those, I am just afraid that if we don't go through RFC, this will just push the discussion / work in an obscure place, and we will lose the diversity of people who comes to read this or are involved in this process. Hence, also why I believe that RFC process is the right place.

I would definitely be able to go around and push my Bootspec v2 by myself and try to ping everyone like I did here, but I think it would reduce the discoverability of the whole ecosystem of work that Bootspec offers to anyone stumbling on it.

Now, maybe, we can have the perspective of potential other members of the RFC steering committee on that?

RaitoBezarius commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+
+Non-x86 systems often rely on device trees to inform firmware or bootloader about the available devices and hardware support. 
+
+There are two distinct requirements for device trees: 
+
+- generic hardware support, where firmware or bootloader selects the appropriate device tree
+- device-specific support, where a hardcoded device tree is required. 
+
+The latter indicates a potentially problematic, non-upstreamed, or in-development platform.
```
This blurb was written while discussing this with @samueldr who well, by their work, is an expert matter in devicetree handling in the Nix ecosystem. 

In my understanding (note that I am not expert matter, and I only know the UEFI side of this), the point I'm trying to make is not the one of bundling or not extra devicetrees but the one of "how do you decide which devicetree to load?"

If your Linux kernel contains all known devicetree, that's awesome news. Though, the problem starts when you don't have the devicetree in the Linux kernel, like I mention it in

> The latter indicates a potentially problematic, non-upstreamed, or in-development platform.

How do you boot a system with a Linux kernel that doesn't know about the right devicetree? I believe @samueldr knows some platforms like those and hence why this is an upstream problem.

The "single devicetree" hardcoding solution is there to offer an escape hatch, not to say this is the way to go.

---

Though, if given those clarifications, what you are saying is that we should always have `FDTDIR` or we should embed all the devicetrees inside the kernel, am I right that you are saying this would cover for all platforms out there in the world, including broken ones?

And therefore, what would be the support of devicetrees in bootspec in that case? Why not just have always the kernel bundle all **known** devicetree? What is the usefulness of having devicetrees expressed in bootspec with what I believe is your proposal — put everything in the kernel, right?

RaitoBezarius replied:
(also as part of open questions, I will reput the link here: https://github.com/aarch64-laptops/edk2/tree/dtbloader-app#dtbloader — which I think is very enlightening on the state of DTB I guess.)

samueldr replied:
These are not "DT" requirements, but the *knowledge NixOS has to have* for ***generation-provided DTs***.

*Thinking about data provided by the generation*, either you want the device tree output following the Linux naming semantics, or you want to pick one. There is no in-between that the boot methods allow.

In the situation where you use the platform firmware provided DT, you just configure neither, because ***it's not data from the generation***. It's ambiant data from the platform firmware.

Reminder, bootspec is not describing the hardware you are running things on, **it describes data from a given NixOS generation that may be required for configuring booting**.

samueldr commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+- `initrdSecrets` will now be a hashmap of strings (key) and strings (value)
+  where the key is the "name" of a secret and the value is the "path" towards a
+  secret.
+- `fdtdir` will be introduced as an opaque string to a directory in a shape as
+  expected by U-Boot extlinux `FDTDIR` directive — it is **optional**
```
Note that while it sounds like we're encoding an U-Boot specific behaviour, in actuality we are encoding the behaviour from the shepherds of the "OSS" device trees. U-Boot only mirrors what the kernel outputs.

Any non-U-Boot system that hopes to load generic device DT *should* follow that de-facto structure, or else it's going to be hitting against more pain.

RaitoBezarius replied:
I tried to write something about this, let me know if you think this is enough, resolving in the meantime.

blitz commented:
@RaitoBezarius Out of curiosity, how would the initrd secrets interact with encryption? Does the user have to encrypt them or do they have to be around in plaintext at bootloader installation time?

RaitoBezarius commented:
> @RaitoBezarius Out of curiosity, how would the initrd secrets interact with encryption? Does the user have to encrypt them or do they have to be around in plaintext at bootloader installation time?

I believe they ought to be around in plaintext at bootloader installation time, but I do have this open question on: should we let systemd decrypt the initrd secrets at switch to configuration time and make them available only for this script execution? Thus, we would have decryption happening only at installation by systemd!

sdht0 commented:
Sharing my thoughts on [XBOOTLDR support](https://uapi-group.org/specifications/specs/boot_loader_specification/) inside bootspec, as per @RaitoBezarius's comments at https://github.com/NixOS/nixpkgs/pull/260241#issuecomment-1789190031, https://github.com/NixOS/nixpkgs/pull/226692#issuecomment-1742817997 and https://github.com/nix-community/lanzaboote/issues/173#issuecomment-1533406338.

In UEFI systems, the EFI system partition (ESP) might have a small size and not easily resizable (e.g., 100MB on a Windows dual-boot). This results in NixOS errors when trying to store kernels and initrd from multiple generations (e.g., [here](https://discourse.nixos.org/t/boot-partition-is-too-small-and-becoming-full/32194) and [here](https://discourse.nixos.org/t/what-to-do-with-a-full-boot-partition/2049)). The solution (apart from trying to resize or relocated the ESP) is use a separate partition for storing the boot files and use the ESP only to install the bootloader.

In NixOS, GRUB [already supports](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/loader/grub/grub.nix#L708) this scenario. `systemd-boot` can also use an extra XBOOTLDR partition, but NixOS does not support configuring such a partition yet (https://github.com/NixOS/nixpkgs/pull/260241).

@RaitoBezarius suggested that XBOOTLDR-type partition should perhaps be explicitly supported inside bootspec as an extension. However, after reading more on bootspec, I'm not sure if XBOOTLDR fits in here. As was pointed out above, bootspec applies to individual generations, while configuring the ESP and XBOOTLDR partitions is a system wide action. Individual generations will not decide which partition its files should be installed to, and enabling the XBOOTLDR partition would mean files from all generations will be moved to that partition. I might be missing something and perhaps @RaitoBezarius can chime in with how bootspec and XBOOTLDR can interact?

Windows unfortunately [still suggests](https://www.reddit.com/r/sysadmin/comments/13yeh3d/windows_11_new_efi_system_partition_requirement/) the ESP size to be 100MB (even Windows updates [fails sometimes](https://support.microsoft.com/en-us/topic/-we-couldn-t-update-system-reserved-partition-error-installing-windows-10-46865f3f-37bb-4c51-c69f-07271b6672ac) because of that), so I think NixOS should generally support the XBOOTLDR partition in it's bootloaders, whether individually in each bootloader or as a policy in something like this bootspec.

sdht0 commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+
+These facts are used as the primary input for bootloader backends like systemd-boot and grub, for creating files in `/boot/loader/entries/` and `grub.cfg`.
+
+In this proposal, we aim to tackle known weaknesses of Bootspec v1, namely:
```
Nit: the list order should reflect the order of the sections below?

RaitoBezarius replied:
Done.

sdht0 commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+- Lack of devicetree
+- Multiple initrds support
+- Rework of initrd secrets mechanism
+- Extra metadata that was discovered to be useful during the year of usage of Bootspec v1
```
This point doesn't seem to have been discussed further. Maybe show an example where it'd be useful?

RaitoBezarius replied:
Removed.

sdht0 commented on rfcs/0164-bootspecv2.md:
```diff
@@ -0,0 +1,300 @@
+integrity.
+
+In practice, initrd secrets are often employed to establish stable fingerprints
+for SSH servers within the initrd, aiding in remote disk decryption on servers.
```
```suggestion
for SSH servers within the initrd, or aiding in remote disk decryption on servers.
```

RaitoBezarius replied:
it's not or, those fingerprints aid the remote disk decryption afaik?

sdht0 replied:
Ah remote decryption yes.

There's also a case for automatically decrypting a server's encrypted disk on boot, without having anything to do with ssh. Atleast that's how I use it.

RaitoBezarius replied:
Yep, agreed.

infinisil commented on rfcs/0164-bootspecv2.md:
Minor: Please change the PR title and file to match the PR number (165)

RaitoBezarius replied:
Done.

tejing1 commented:
I may be misunderstanding here, but from what I can see, the bootloader backend's responsibility is totally unspecified when it comes to initrd secrets.

Is this proposal assuming that the nix code creating a generation and the bootloader backend are just magically in agreement about this? Does no one ever switch to a different bootloader backend on an existing system with existing generations?

What, precisely, is the bootloader backend required to ensure about the stage1 environment when boot.json contains a non-empty `initrdSecrets` value? I think this needs to be specified clearly, to the point that, in particular, switching to a different bootloader backend can be done and have previously existing generations actually function on the new backend.

RaitoBezarius commented:
> I may be misunderstanding here, but from what I can see, the bootloader backend's responsibility is totally unspecified when it comes to initrd secrets.
> 
> Is this proposal assuming that the nix code creating a generation and the bootloader backend are just magically in agreement about this? Does no one ever switch to a different bootloader backend on an existing system with existing generations?
> 
> What, precisely, is the bootloader backend required to ensure about the stage1 environment when boot.json contains a non-empty `initrdSecrets` value? I think this needs to be specified clearly, to the point that, in particular, switching to a different bootloader backend can be done and have previously existing generations actually function on the new backend.

We can specify this but I surmise this will be relying a lot on systemd stage 1 semantics anyway.

kevincox commented on rfcs/0165-bootspecv2.md:
We have a bunch of proposed nominations from @RaitoBezarius. I would appreciate if the nominees could accept or deny.

- @lheckemann 
- @GovanifY
- @06kellyjac (assumed typo, 👍 on the comment, please explicitly accept)
- @hesiod
- @amjoseph-nixpkgs

Just a reminder that there is no special qualification required to be a shepherd. The role just involves ensuring the RFC is moving forward by collecting feedback, ensuring the RFC text evolves to incorporate any raised concerns and eventually moving towards FCP. (The details of the process are documented in the [README](https://github.com/NixOS/rfcs/blob/master/README.md)).

RaitoBezarius replied:
@lheckemann do you confirm to be co-author my dear linus? :p -- OK for shepherding, thank you!

infinisil replied:
Speaking as a member of the RFC steering committee, after people [confirmed their nominations](https://github.com/NixOS/rfcs/pull/165#issuecomment-1855690878), we approve of the shepherd team as is in the RFC:
- Leader: @JulienMalka
- Shepherds: @lheckemann (or co-author) @GovanifY @06kellyjac @hesiod

JulienMalka commented:
If there is a need for more shepherds, I'd like to self nominate for that. I've been following this RFC carefully, worked with bootspec v1 implementation in the `systemd-boot` backend (see https://github.com/NixOS/nixpkgs/pull/263442), and have been working with initrd secrets, so I am interested in the subject.

RaitoBezarius commented:
@infinisil @amjoseph-nixpkgs I addressed the concern over the RFC being not interesting enough for this repo. I will ask you that we will at least mandate in this RFC the move to another repository.

@grahamc @cole-h Do you have any proposal for moving the RFC in another repository? I think either nixpkgs or nix-community or nixos would be the right place to do it. I also addressed various feedback on the initrd secrets and transition period.

@sdht0 I addressed your comments.

@tejing1 I addressed your comments.

I added @JulienMalka as a shepherd leader (after a IRL discussion). I am still waiting feedback from the other potential nominated shepherds.

tejing1 commented:
Much clearer regarding the initrd secrets. One thing could maybe do with a bit more clarification, though: Are the key strings of the initrd secrets hashmap meaningful to the bootloader backend's responsibilities at all? My inference from reading it is that they aren't, but it's a little vague.

samueldr commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,372 @@
+  Bootspec which should be an hashmap of strings (key) and strings (value)
+  where the key is the "name" of a secret and the value is the **filesystem path** towards a
+  secret.
+- `fdtdir` will be introduced as an opaque string to a directory in a shape as
+  the kernel outputs give them, e.g. U-Boot's extlinux directive `FDTDIR` is an
+  example of this behavior — it is **optional**
```
```suggestion
- `fdtdir` will be introduced as an opaque string to a directory following the structure
   of the `dtbs` folder from the kernel, e.g. U-Boot's extlinux directive `FDTDIR` is an
  example of this behavior — it is **optional**
```

Alternatively to the `dtbs` folder, it could be phrased referencing the fact it's the `dtbs_install` target of the build.

lheckemann commented:
@govanify @06kellyjac @hesiod @amjoseph-nixpkgs Could you confirm whether you accept your nominations as shepherds for this RFC?

06kellyjac commented:
Yes I accept

GovanifY commented:
I accept, sorry for the delay have been busier than anticipated!

hesiod commented:
Thanks for the nomination, I accept.

RaitoBezarius commented:
I updated everyone status on the specification. Please take the time to review the RFC text again, I will soon re-address the new comments. (I prefer to do it in batch.)

Once we are done with that, maybe, we can setup a quick call @JulienMalka to sync everyone on that RFC and see how we want to proceed.

I am also planning to do a prototype implementation before the RFC as always.

Luflosi commented:
When using flakes, one can add `{ system.configurationRevision = self.rev or "dirty"; }` to their configuration. This allows figuring out which git commit of the flake was used to build the system by executing `nixos-version --configuration-revision` or `nixos-rebuild list-generations`.
Can we add a field for the content of `system.configurationRevision`?
Bootloaders could then also show this information in a useful way. Currently there is no easy way for a bootloader to show this information as the only place where it is currently stored is by being hardcoded into the `nixos-version` shell script.

Maybe this is already covered by somehow adding the information to the "label"?

RaitoBezarius commented:
> When using flakes, one can add `{ system.configurationRevision = self.rev or "dirty"; }` to their configuration. This allows figuring out which git commit of the flake was used to build the system by executing `nixos-version --configuration-revision` or `nixos-rebuild list-generations`. Can we add a field for the content of `system.configurationRevision`? Bootloaders could then also show this information in a useful way. Currently there is no easy way for a bootloader to show this information as the only place where it is currently stored is by being hardcoded into the `nixos-version` shell script.
> 
> Maybe this is already covered by somehow adding the information to the "label"?

This is already covered by the label information IMHO, bootloaders decide what is the title and what information they need to encode for the label. Adding configuration revision seems out of scope just for cosmetic purpose.

06kellyjac commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+author: Ryan Lahfa (@raitobezarius)
+co-authors: Linus Heckemann (@lheckemann)
+shepherd-team:
+- @06kellyjack
```
```suggestion
- @06kellyjac
```

JulienMalka commented:
@RaitoBezarius @GovanifY  @06kellyjac @hesiod @lheckemann  let's set up a first sync call either the week of the 14th of the week of the 21st. Can you please fill [this when2meet](https://www.when2meet.com/?22974581-3SlPH) so that we can find an appropriate schedule for this? I think a 45mn call should be more than enough.

nikstur commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+    // (Required) The label of the system. It should contain the operating system, kernel version,
+    // and other user-relevant information to identify the system. This corresponds
+    // loosely to `config.system.nixos.label`.
+    "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",
```
I'd like this to be structured information, i.e. make label an attrset that contains `distroName, `release`, `uname` etc. This is because ultimately it is the bootloader builder's resposibility to assemble this information so that it can be displayed nicely using the facilities the specific bootloader provdes. What is "nice" might also change between bootloaders.

We have this problem in Lanzaboote already where we have a suboptimal display name for a generation because we have to just use the label as is.

samueldr replied:
I believe this is answered here:

 - https://github.com/NixOS/rfcs/pull/165#issuecomment-1859188215

* * *

My understanding here is that providing structured data is undesirable as this represents a snapshot of the already serialized boot entry configuration. It is not intended to be re-parsed for further processing.

In this instance, it is the [`system.nixos.label`](https://search.nixos.org/options?channel=unstable&show=system.nixos.label&from=0&size=50&sort=alpha_asc&type=packages&query=++++system.nixos.label) configuration, which is an arbitrary label that can be configured by the end-user.

What, really, I'm wondering is: is the comment describing it correct? Shouldn't it be just `system.nixos.label`, and if not, why?

boomshroom replied:
Generally the bootloader would use the generation number, except that the generation number is nowhere to be seen in the bootspec, and for good reason, since it's the one thing that's guaranteed to change between rebuilds, while the boot.json resides within the system path itself, which is only supposed to depend on what's explicitly provided to nix at build-time. Bootspec doesn't even provide _any_ information that other generations even exist beyond specialisations for that particular build.

So bootloaders are already forced to look elsewhere than just the provided system path in order to generate its boot entries. At that point, why not just take a peek at /etc/os-release to get more granular information and reduce some of the duplication? If the boot menu has any kind of hierarchy, then saying "NixOS" for every single subentry seems unnecessary when the version information is all you need. The only information in `label` that isn't in `os-release` is the kernel version.

RaitoBezarius replied:
Any bootloader implementation can peek at the `/etc/os-release` as long as it's determined at build-time and generated an appropriate label.

Furthermore, there's no reason to make `/etc/os-release` a dependency, someone could use bootspec on systems without it and with a special bootloader implementation making use of another knowledge.

The kernel version is also known at build-time. There's no need to literally read `/etc/os-release` per se.

Ideally, the bootloader should try to honor as much as possible the `label` or come up with better generation of the `label`, sure, in lanzaboote, we have a problem that is, we need better tools to format the title because we can run in various excessive cleverness from systemd-boot that forces us to assume something about the label itself and manipulate it. As soon someone overrides forcibly that label, we are doomed.

But customization of boot title entry is a complicated matter, one, that exceeds the scope of bootspec maybe…

To put in another way, there's a tension between "bootspec's label is an opaque string that must be honored" and "my bootloader is doing weird things to the sorting order of the label / eating some words / I would like to show a 2-level menu and I don't know where to split the label string".

The resolution of this situation might be to have `preferredLabel` and the data @nikstur proposes to have which provides the preferred label in a structured way which the bootloader implementation may decide to reassemble for limitations of the software or to better accommodate the UI. Though, this increases the complexity of the design and opens a can of worms.

JulienMalka commented:
@06kellyjac @hesiod I invited you both to a matrix room so that we can synchronously converge towards dates for our meeting, could you join please ?

lheckemann commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+  // to securely make available the secret inside the initrd phase.
+  // This may involve leveraging TPM2 via systemd-credentials or any measure you deem
+  // to be reasonable in the context.
+  // The legacy behavior is to prepare a CPIO archive for each file and
+  // extend the `initrds` fields with those CPIO archives.
```
```suggestion
  // The legacy behavior is to prepare a CPIO archive for each file and
  // concatenate these with the initrds from the `initrds` field to form
  // a single initrd for boot. This exposes the secrets to anyone who can
  // read the boot partition, and is discouraged in favour of mechanisms
  // which provide stronger protection for the secrets.
```

lheckemann commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+  // The legacy behavior is to prepare a CPIO archive for each file and
+  // extend the `initrds` fields with those CPIO archives.
+  // Make sure the location where the secrets are dropped in the initrd are visible
+  // for the user.
```
Unclear, what does this mean?

RaitoBezarius replied:
Imagine you are activating the script in a rescue system and the secrets are in `/mnt/etc/secrets/...` instead of `/etc/secrets/...` and you are not fully chrooted inside `/mnt` or whatever, the location won't be visible at assembly time.

This sort of problem cause a lot of difficult to understand to most users because activation now depends on an invisible files to assemble bootables (arguably, this is not the only mechanism like this we have.).

How would you reframe this concept?

RaitoBezarius commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+Moreover, an additional initrd is frequently used to store CPU microcode. To
+ensure compatibility and flexibility, it is essential to rework the initrd
+support in the Boot Loader Specification. The proposed changes aim to treat
+initrds as a set, allowing bootloaders to handle a list of initrds or a single
```
That's correct but I guess I should add some remarks on what is possible in terms of ordering.
For example, CPU microcode must always come first, uncompressed.

lheckemann replied:
That makes it all the more important :D I was mostly thinking of precedence with colliding paths, but microcode-first is obviously especially important. I'm not sure we even need to detail it that much in the text, but we shouldn't call it a "set" :)

lheckemann replied:
```suggestion
initrds as a list, allowing bootloaders to handle a list of initrds or a single
```

RaitoBezarius commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+- Improve non-x86 support in Bootspec, emphasizing the importance of
+  devicetrees.
+- Address and reduce the risks associated with initrd secrets, providing a
+  default "secure" implementation within the systemd ecosystem and offering
```
My current feeling about scripted initrd is that we will start deprecation in 24.05, thus, getting the nice features for it seems to be a waste of energy, but I may be wrong (?).

We can keep v1 working as-is for scripted initrd as we go in deprecation of v1 too, no?

RaitoBezarius commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+In practice, initrd secrets are often employed to establish stable fingerprints
+for SSH servers within the initrd, or aiding in remote disk decryption on servers.
+To address these issues, this RFC proposes moving away from the "appender
+script" model in Bootspec v1 and instead adopting a hash map format to
```
Sure!

RaitoBezarius commented on rfcs/0165-bootspecv2.md:
```diff
@@ -0,0 +1,375 @@
+  //  - In particular, there is no expectation that such nested specialisations will be handled by consumers of bootspec documents.
+  //  - Each specialisation document may contain arbitrary further keys (extensions), like the top-level document.
+  //  - The semantics of these should be the same as when these keys are used at the top level, but only apply for the given specialisation.
+  "org.nixos.specialisation.v2": {
```
Well, each specialization must have a valid v2 bootspec document, so it's a change in the expectations of the v1 specialization specification, so it deserves a bump (?).

(I also like bumps to test fully our capabilities to avoid ossification though.)

samueldr commented:
(I'm not a shepherd... Though I can be pinged/asked questions about boot given my domain knowledge.)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-03-05/40851/1

adamcstephens commented:
> During the meeting, we also decided that ideally, some implementation work should be done before the RFC get merged, mainly updating https://github.com/DeterminateSystems/bootspec and the NixOS module.

Should this repository be moved under the NixOS organization?

jonringer commented:
Hey shepherds, even if implementation is still ongoing, you can move this into FCP. The implementation doesn't need to be completed before this RFC has been merged.

Seems like there's agreement on the direction of this RFC; unless there are still outstanding concerns, the RFCSC recommendation would be to move to FCP.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-03-19/41829/1

RaitoBezarius commented:
The FCP cannot happen before a first implementation is done in the
ecosystem, IMHO.

Le mar. 19 mars 2024, 09:02, nixos-discourse ***@***.***> a
écrit :

> This pull request has been mentioned on *NixOS Discourse*. There might be
> relevant details there:
>
> https://discourse.nixos.org/t/rfcsc-meeting-2024-03-19/41829/1
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/rfcs/pull/165#issuecomment-2007574071>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACMZRFWDPFP7PE3G7WGV4TYZBORRAVCNFSM6AAAAAA6ZOZO42VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMBXGU3TIMBXGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-04-02/42643/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-04-16/43512/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-05-14/45414/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-05-28/46113/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-06-10/46817/1

MMesch commented:
RFCSC: Since there's no progress, we'll mark this as a draft. Feel free to undraft at any time when you're continuing with it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfcsc-meeting-2024-06-24/47589/1

Prince213 commented:
Can we have an additional section on UKI images?  Currently we have NixOS modules for UKI, but it doesn't show up in bootspec.

Related issue: https://github.com/NixOS/nixpkgs/issues/413039

RaitoBezarius commented:
Thanks for your interest in this RFC. I am not planning to drive any large scale changes in Nixpkgs for the time being. Feel free to reopen it if you are interested.
